src/main/java/cd/go/artifact/docker/registry/AWSTokenRequestGenerator.java

```diff
@@ -0,0 +1,75 @@
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.auth.DefaultAWSCredentialsProviderChain;
+import com.amazonaws.client.builder.AwsSyncClientBuilder;
+import com.amazonaws.services.ecr.AmazonECR;
+import com.amazonaws.services.ecr.AmazonECRClient;
+import com.amazonaws.services.ecr.AmazonECRClientBuilder;
+import com.amazonaws.services.ecr.model.GetAuthorizationTokenRequest;
+import com.amazonaws.services.ecr.model.GetAuthorizationTokenResult;
+import com.thoughtworks.go.plugin.api.logging.Logger;
+
+import java.util.Base64;
+
+import static org.apache.commons.lang.StringUtils.isNotBlank;
+
+public class AWSTokenRequestGenerator {
+    public static final Logger LOG = Logger.getLoggerFor(AWSTokenRequestGenerator.class);
+    private AwsSyncClientBuilder<AmazonECRClientBuilder, AmazonECR> builder;
+
+    AWSTokenRequestGenerator() {
+        builder = AmazonECRClient.builder();
+    }
+
+    AWSTokenRequestGenerator(AwsSyncClientBuilder<AmazonECRClientBuilder, AmazonECR> builder) {
+        this.builder = builder;
+    }
+
+    AwsSyncClientBuilder<AmazonECRClientBuilder, AmazonECR> getBuilder() {
+        return builder;
+    }
+
+    String[] getUsernameAndPasswordFromECRToken(ArtifactStoreConfig artifactStoreConfig) {
+        builder.setRegion(artifactStoreConfig.getAwsRegion());
+        setCredentialsProvider(artifactStoreConfig);
+        GetAuthorizationTokenResult authorizationTokenResult = builder.build().getAuthorizationToken(new GetAuthorizationTokenRequest().withRegistryIds(getAwsAccountIdFromURL(artifactStoreConfig.getRegistryUrl())));
+        String authorizationToken = authorizationTokenResult.getAuthorizationData().get(0).getAuthorizationToken();
+        return new String(Base64.getDecoder().decode(authorizationToken)).split(":");
+    }
+
+    void setCredentialsProvider(ArtifactStoreConfig artifactStoreConfig) {
+        if (isNotBlank(artifactStoreConfig.getAwsAccessKeyId()) || isNotBlank(artifactStoreConfig.getAwsSecretAccessKey())) {
+            AWSStaticCredentialsProvider awsStaticCredentialsProvider = new AWSStaticCredentialsProvider(new BasicAWSCredentials(artifactStoreConfig.getAwsAccessKeyId(), artifactStoreConfig.getAwsSecretAccessKey()));
+            builder.setCredentials(awsStaticCredentialsProvider);
+            LOG.debug("Using the AWS keys configured in the specified artifact store.");
+        }
+        else {
+            builder.setCredentials(new DefaultAWSCredentialsProviderChain());
+            LOG.debug("Setting the default aws credentials chain provider. This happens if the specified artifact store does not have aws keys configured. The default chain checks for environment variables, system properties, credentials profile, instance profile.");
+        }
+    }
+
+    private String getAwsAccountIdFromURL(String registryUrl) {
+        return registryUrl.split("//")[1].split("\\.")[0];
```

Comment:
Check if `registryUrl` is specified with the uri scheme. Otherwise, it will result in an error.  Please, can you also update the UI with acceptable ECR registry URL formats?

Reply:
> can you also update the UI with acceptable ECR registry URL formats

You mean, provide help text with examples for ecr, dockerhub, gcr for the registry url input in the UI -  "https://aws_account_id.dkr.ecr.us-east-1.amazonaws.com", "https://index.docker.io/v1/", "gcr.io/gocd-server/" ?

Reply:
Currently, there is no validation on a registry URL. So a user can specify it as `https://aws_account_id.dkr.ecr.us-east-1.amazonaws.com` or `aws_account_id.dkr.ecr.us-east-1.amazonaws.com`. I initially specified as `aws_account_id.dkr.ecr.us-east-1.amazonaws.com` without `https://`. I copied it from the AWS ecr console and assumed that is the right format. Later it fails with `ArrayIndexOutOfBoundException`.

I am thinking that having artifact store config with the following UI makes more sense -

- Select the registry type from `ECR, other` first.
- If the selected registry is ECR then show form with -
      - RegistryID along with `AWS Access Key Id`, `AWS Secret Key` and `Region`. Since you are providing on ECR registry id there is no need of `getAwsAccountIdFromURL`
- If the selected registry is `Other` then show form with -
      - RegistryURL along with other fields

@varshavaradarajan - WDYT?